feat: add systemd-machined integration and SSH over VSOCK

elsbrock opened the pull request:
Adds systemd-machined integration and SSH over VSOCK support for MicroVMs.

## machined integration

`microvm.registerWithMachined = true` registers VMs with machined:

```
$ machinectl list
MACHINE    CLASS SERVICE     OS VERSION ADDRESSES
authelia   vm    microvm.nix -  -       -
blackpearl vm    microvm.nix -  -       -
```

- `machinectl list/status/show` for visibility
- `machinectl terminate/kill` to stop VMs

This PR prefers `RegisterMachineEx` on systemd 259+ and falls back to legacy machined calls on older hosts.  
When available, it registers `VSockCID` / `SSHAddress` metadata (plus network interfaces).

`machinectl reboot/poweroff/terminate/kill` all stop the VM without auto-restart (systemd treats signals from machined as intentional stops). Use `systemctl restart microvm@<name>` for restarts.

## SSH over VSOCK

`microvm.vsock.ssh.enable = true` configures guest sshd to listen on VSOCK (requires `microvm.vsock.cid`):

```
$ microvm -s myvm
Connecting to myvm via VSOCK...
[myvm]$
```

Works with qemu/crosvm/kvmtool (native AF_VSOCK) and cloud-hypervisor (socket mux).

This PR also updates `microvm -s` argument handling so SSH options can be forwarded via `--`, e.g.:

- `microvm -s myvm -- -l root -i /path/to/key`

## Other changes

- `microvm -l` respects configured `microvm.stateDir`
- `microvm.machineId` option for consistent UUID handling across machined/SMBIOS/guest machine-id use
- QEMU shutdown script handles missing control socket (fixes restart after external kill)
- imperative `microvm@%i` template tolerates missing `microvm-register` / `microvm-unregister` scripts

## Tests

- expands machined test coverage for VM metadata (`VSockCID`, `SSHAddress`) where applicable
- adds a focused `microvm -s` argument-handling test
- adds imperative template test for start/stop without register scripts

Closes #123

SuperSandro2000 commented on lib/runner.nix:
```diff
@@ -28,6 +28,88 @@ let
+  machineUuid =
+    if microvmConfig.machineId != null
+    then microvmConfig.machineId
+    else builtins.readFile (vmHostPackages.runCommand "machine-uuid" {} ''
```
We should try to avoid IFD as much as possible.

SuperSandro2000 commented on lib/runner.nix:
```diff
@@ -28,6 +28,88 @@ let
+
+    ${if tapInterfaceNames == [] then ''
+    # No TAP interfaces, use simple RegisterMachine
+    ${vmHostPackages.systemd}/bin/busctl call \
```
```suggestion
    '' + if tapInterfaceNames == [] then ''
    # No TAP interfaces, use simple RegisterMachine
    busctl call \
```

SuperSandro2000 commented on lib/runner.nix:
```diff
@@ -28,6 +28,88 @@ let
+        "$MACHINE_NAME" \
+        16 $UUID_BYTES \
+        "microvm.nix" \
+        "vm" \
```
It would probably be easier to read, if we don't split every word onto a new line.

SuperSandro2000 commented on lib/runner.nix:
```diff
@@ -79,24 +79,27 @@ let
     if [ -e /sys/class/net/${name}/ifindex ]; then
-      IFINDICES="$IFINDICES $(cat /sys/class/net/${name}/ifindex)"
+      IFINDEX=$(cat /sys/class/net/${name}/ifindex)
+      IFINDICES="$IFINDICES $IFINDEX 0"
```
Why do we change those lines in `` fix: correct D-Bus signature for RegisterMachineWithNetwork`` again? Please try to keep a clean history even if it is less commits.

SuperSandro2000 commented on nixos-modules/microvm/vsock-ssh.nix:
```diff
@@ -0,0 +1,42 @@
+    '';
+  };
+
+  config = lib.mkIf (cfg.ssh.enable && cfg.cid != null) {
```
```suggestion
  config = lib.mkIf cfg.ssh.enable {
```
Otherwise, the assertion below does never trigger.

SuperSandro2000 commented on nixos-modules/microvm/vsock-ssh.nix:
```diff
@@ -0,0 +1,42 @@
+
+    services.openssh = {
+      enable = true;
+      startWhenNeeded = true;
```
Lets not change that for most people and keep the ssh defaults.

SuperSandro2000 commented on nixos-modules/microvm/vsock-ssh.nix:
```diff
@@ -0,0 +1,42 @@
+      startWhenNeeded = true;
+      # Don't listen on network interfaces by default when using VSOCK
+      # Users can override this if they want both
+      listenAddresses = lib.mkDefault [];
```
This would break the microvms of most users I know. Speaks there anything against keeping the defaults?

SuperSandro2000 commented on nixos-modules/microvm/options.nix:
same for ` docs: document machinectl limitations for VMs `

elsbrock commented on nixos-modules/microvm/default.nix:
```diff
@@ -35,5 +35,12 @@ in
+    # This ensures the guest machine-id matches the UUID passed to machined and SMBIOS
+    environment.etc."machine-id" = lib.mkIf (config.microvm.machineId != null) {
+      mode = "0444";
+      text = builtins.replaceStrings ["-"] [""] config.microvm.machineId + "\n";
```
Hmm, that doesn't exist in my `nix`

SuperSandro2000 replied:
I think it is new with 25.11.

elsbrock committed:
feat: add SSH over VSOCK support

Enable SSH access to guests via VSOCK without network configuration.

- New option: microvm.vsock.ssh.enable
- New command: microvm -s <vmname> for VSOCK SSH
- Supports qemu/crosvm/kvmtool (AF_VSOCK) and cloud-hypervisor (socket mux)
- Guest uses systemd's ssh-generator for automatic sshd-vsock.socket

astro commented on nixos-modules/host/default.nix:
```diff
@@ -113,6 +113,19 @@ in
+        ];
+        serviceConfig.ExecStopPost = lib.optionals runner.registerWithMachined [
+          "+${stateDir}/${name}/current/bin/microvm-unregister"
+        ];
```
This covers only the declarative case where a discete `microvm@${hostName}` unit is generated.

We also want to support imperative mode with the `microvm@` template unit. Can we somehow call `microvm-{un,}register` at runtime, without systemd panicking if it doesn't exist when the feature is disabled?

astro committed:
nixos-modules/microvm/system: set a default networking.hostId like systemd would do

SuperSandro2000 commented on nixos-modules/microvm/system.nix:
```diff
@@ -58,5 +58,14 @@
+    # Set /etc/machine-id from machineId if provided
+    # This ensures the guest machine-id matches the UUID passed to machined and SMBIOS
+    environment.etc."machine-id" = lib.mkIf (config.microvm.machineId != null) {
+      text = builtins.replaceStrings ["-"] [""] config.microvm.machineId + "\n";
```
```suggestion
      text = lib.replaceString "-" "" config.microvm.machineId + "\n";
```

elsbrock committed:
Filter checks to not contain darwin only checks on Linux and vice versa

elsbrock committed:
feat: register MicroVMs with systemd-machined

Add optional integration with systemd-machined for VM visibility and management.

- New option: microvm.registerWithMachined (default false)
- New option: microvm.machineId for consistent UUID across machined/SMBIOS/guest
- Shared makeMachineUuid function in lib to avoid duplication
- QEMU sets SMBIOS UUID for guest visibility
- Host module registers VM after hypervisor starts
- Graceful shutdown handling when VM is killed via machinectl
- NixOS test for machined integration

elsbrock committed:
fix: use RegisterMachineEx fallback path and add targeted machined regressions

SuperSandro2000 commented on checks/machined.nix:
```diff
@@ -0,0 +1,77 @@
+            ${lib.optionalString (hypervisor != "cloud-hypervisor") ''
+              # On systemd >=259 RegisterMachineEx path should expose VSOCK/SSH metadata
+              host.succeed("machinectl show '${vmName}' --property=VSockCID | grep -q 'VSockCID=${toString vsockCid}'")
+              host.succeed("machinectl show '${vmName}' --property=SSHAddress | grep -q 'SSHAddress=vsock/${toString vsockCid}'")
```
Haha, not a concern for NixOS

SuperSandro2000 committed:
flake.lock: Update

Flake lock file updates:

• Updated input 'nixpkgs':
    'github:nixos/nixpkgs/7df7ff7d8e00218376575f0acdcc5d66741351ee?narHash=sha256-gTrEEp5gEspIcCOx9PD8kMaF1iEmfBcTbO0Jag2QhQs%3D' (2025-10-02)
  → 'github:nixos/nixpkgs/aca4d95fce4914b3892661bcb80b8087293536c6?narHash=sha256-E1bxHxNKfDoQUuvriG71%2Bf%2Bs/NT0qWkImXsYZNFFfCs%3D' (2026-03-06)
• Updated input 'spectrum':
    'git+https://spectrum-os.org/git/spectrum?ref=refs/heads/main&rev=c5d5786d3dc938af0b279c542d1e43bce381b4b9' (2025-10-03)
  → 'git+https://spectrum-os.org/git/spectrum?ref=refs/heads/main&rev=fe39e122d898f66e89ffa17d4f4209989ccb5358' (2026-02-27)

SuperSandro2000 committed:
machined: before

SuperSandro2000 committed:
Fix machined registration with systemd 259

SuperSandro2000 committed:
Fix doc build

oddlama commented on nixos-modules/microvm/system.nix:
```diff
@@ -58,5 +58,14 @@
+      text = lib.replaceString "-" "" config.microvm.machineId + "\n";
+    };
+    # Generate hostId from machine-id like systemd would do
+    networking.hostId = lib.mkDefault (
```
if a user sets `machineId = null` this will error out. should probably be guarded with mkIf machineId != null

SuperSandro2000 replied:
I've opened #499, but I only eye balled this, so it could be another issue like this in the code base.

SuperSandro2000 commented on nixos-modules/microvm/options.nix:
```diff
@@ -486,6 +486,53 @@ in
+
+    machineId = mkOption {
+      type = with types; nullOr str;
+      default =
```
I do not have a good idea how to make this migration smooth unless we introduce a state version.
